File: a1/plotimages.m
function plotimages(images, Y ,scale,proportion)
% function plotimages(images, Y, scale, proportion)
%
% images = images, must be in a 3-dimensional matrix (x by y by n)
% for example if X is 64 by 400 and size of each image is 8 by 8, images=reshape(X,8,8,400);
%
% Y = where to plot the image (Y(1,:) by Y(2,:))
%
% proportion = proportion of the data to be ploted (proportion <= 1).
% for example if there are 400 data points proportion = 1, plots
% all 400 data points and proportion = 0.5 plot only 200 data points 
% (i.e. 1th, 3th, 5th, ...)
% Ali Ghodsi 2006

Y=normr(Y);
inc=floor(1/proportion);
% scale = scale of each image wrt to figure size (scale<1)
%scale=10;
xoff=0;
yoff=0;

xlim = get(gca, 'XLim');
ylim = get(gca, 'YLim');


width = (xlim(2) - xlim(1)) * scale;
height = (ylim(2) - ylim(1)) * scale;

colormap(gray);

image_width = size(images,1);
image_height = size(images,2);
n_images = size(images,3);

xs = Y(1,:) + xoff;
ys = Y(2,:) + yoff;
hold on
for counter = 1:inc:n_images
   
	current_image = 1-reshape(images(:,:,counter), [image_width image_height]);
	imagesc( ...
		[xs(counter)		xs(counter)+width], ...
		[ys(counter)+height	ys(counter)], ...
		current_image' ...
	);
end
xlabel ('x')
ylabel ('y')
hold off

File: a1/plotimages.py
import matplotlib.pyplot as plt
from matplotlib.offsetbox import AnnotationBbox, OffsetImage
import numpy as np
from numpy.linalg import norm

def plotimages(images, Y, scale=0.9, proportion=0.2):
    '''
    Input:
        images: images, must be in a 3-dimensional matrix (x by y by n)
            e.g. if X is 64 by 400 and size of each image is 8 by 8, images=np.reshape(X,(8,8,400));
        Y: coordinates of where to plot the image (Y(1,:) by Y(2,:)
        scale: scale of each image wrt to figure size (scale <= 1, e.g. 0.8)
        proportion: proportion of the data to be ploted (proportion <= 1)
            e.g. if there are 400 data points proportion = 1, plots
            all 400 data points and proportion = 0.5 plot only 200 data points 
            (i.e. 1th, 3th, 5th, ...)
    Ali Ghodsi 2006
    '''
    Y /= norm(Y, axis = 1, ord=2).reshape((Y.shape[0],1))
    inc = np.floor(1/proportion)
    
    image_width = images.shape[0]
    image_height = images.shape[1] 
    n_images = images.shape[2]
    
    plt.gray()
    fig = plt.gcf()
    fig.clf()
    ax = plt.subplot(111)
    ax.set_xlim((Y[0,:].min()*1.2,Y[0,:].max()*1.2))
    ax.set_ylim((Y[1,:].min()*1.2,Y[1,:].max()*1.2))
    
    for counter in np.arange(0,n_images,inc):
        counter = int(counter)
        xy = (Y[0,counter],Y[1,counter])
        current_image = 1-np.reshape(images[:,:,counter], (image_width,image_height))
        imagebox = OffsetImage(current_image, zoom=scale)
        ab = AnnotationBbox(imagebox, xy, xybox=(1., -1.), xycoords='data', boxcoords="offset points", frameon=False)                                  
        ax.add_artist(ab)
        plt.draw()
    #plt.show()
